snuba/query/processors/arrayjoin_optimizer.py

```diff
@@ -0,0 +1,288 @@
+
+    return cond_tags_key | having_tags_key
+
+
+class ArrayjoinOptimizer(QueryProcessor):
+    """
+    Apply two optimizations to reduce the performance impact of tags_key
+    and tags_value expressions that are sent to clickhouse as
+    arrayJoin(tags.key)
+
+    - it performs the arrayJoin only once in case both tags_key and
+        tags_value are present in the query by arrayJoining on the
+        key, value pairs instead of arrayJoining twice, once on keys
+        and the second on values.
+    - prefilter the keys with arrayFilter inside the arrayJoin if there
+        are conditions in the query that specify which tags we want
+        instead of doing the entire arrayJoin and filter the results.
+    """
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        arrayjoin_pattern = FunctionCall(
+            None,
+            String("arrayJoin"),
+            (
+                Column(
+                    None,
+                    None,
+                    Param("col", Or([String("tags.key"), String("tags.value")])),
+                ),
+            ),
+        )
+
+        arrayjoins_in_query = set()
+        for e in query.get_all_expressions():
+            match = arrayjoin_pattern.match(e)
+            if match is not None:
+                arrayjoins_in_query.add(match.string("col"))
+
+        filtered_tags = [
+            LiteralExpr(None, tag_name) for tag_name in get_filtered_tag_keys(query)
+        ]
+
+        def replace_expression(expr: Expression) -> Expression:
+            """
+            Applies the appropriate optimization on a single arrayJoin expression.
+            """
+            match = arrayjoin_pattern.match(expr)
+            if match is None:
+                return expr
+
+            if arrayjoins_in_query == {"tags.key", "tags.value"}:
+                # Both tags_key and tags_value expressions present int the query.
+                # do the arrayJoin on key-value pairs instead of independent
+                # arrayjoin for keys and values.
+                array_pos = (
+                    LiteralExpr(None, 1)
+                    if match.string("col") == "tags.key"
+                    else LiteralExpr(None, 2)
+                )
+
+                if not filtered_tags:
+                    return _unfiltered_tag_pairs(expr.alias, array_pos)
+                else:
+                    return _filtered_tag_pairs(expr.alias, filtered_tags, array_pos)
+
+            elif filtered_tags:
+                # Only one between tags_key and tags_value is present, and
+                # it is tags_key since we found filtered tag keys.
+                return _filtered_tag_keys(expr.alias, filtered_tags)
+            else:
+                # No viable optimization
+                return expr
+
+        query.transform_expressions(replace_expression)
+
+
+def _unfiltered_tag_pairs(alias: Optional[str], array_pos: LiteralExpr) -> Expression:
```

Comment:
all the methods below are introduced to make the expressions simpler to understand and disect even if some times they are used only once.
Expressions like this:
```
(arrayJoin(arrayFilter(
    #       pair -> arrayElement(pair, 1) IN (tags),
    #       arrayMap((x,y) -> [x,y], tags.key, tags.value)
    #  )) as all_tags)[1]
```
in the ast are gigantic and they can get to 20 levels of nesting, which makes them very hard to understand.